sw/device/tests/pmod/i2c_host_eeprom_test.c

```diff
@@ -118,37 +119,46 @@ static status_t write_read_page(void) {
   return OK_STATUS();
 }
 
-static status_t test_init(void) {
+static status_t test_init(dif_i2c_t *i2c, uint8_t i2c_instance,
+                          i2c_pinmux_platform_id_t platform) {
+  const uintptr_t kI2cBaseAddrTable[] = {TOP_EARLGREY_I2C0_BASE_ADDR,
+                                         TOP_EARLGREY_I2C1_BASE_ADDR,
+                                         TOP_EARLGREY_I2C2_BASE_ADDR};
+  TRY_CHECK(i2c_instance < ARRAYSIZE(kI2cBaseAddrTable));
+
+  dif_pinmux_t pinmux;
   mmio_region_t base_addr =
       mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR);
 
-  TRY(dif_rv_core_ibex_init(base_addr, &rv_core_ibex));
-
-  base_addr = mmio_region_from_addr(TOP_EARLGREY_I2C2_BASE_ADDR);
-  TRY(dif_i2c_init(base_addr, &i2c));
+  base_addr = mmio_region_from_addr(kI2cBaseAddrTable[i2c_instance]);
+  TRY(dif_i2c_init(base_addr, i2c));
 
   base_addr = mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR);
   TRY(dif_pinmux_init(base_addr, &pinmux));
 
-  TRY(i2c_testutils_select_pinmux(&pinmux, 2));
+  TRY(i2c_testutils_select_pinmux(&pinmux, i2c_instance, platform));
 
-  TRY(dif_i2c_host_set_enabled(&i2c, kDifToggleEnabled));
+  TRY(dif_i2c_host_set_enabled(i2c, kDifToggleEnabled));
 
   return OK_STATUS();
 }
 
 bool test_main(void) {
-  status_t test_result;
-  CHECK_STATUS_OK(test_init());
-
-  dif_i2c_speed_t speeds[] = {kDifI2cSpeedStandard, kDifI2cSpeedFast,
-                              kDifI2cSpeedFastPlus};
-
-  test_result = OK_STATUS();
-  for (size_t i = 0; i < ARRAYSIZE(speeds); ++i) {
-    CHECK_STATUS_OK(i2c_testutils_set_speed(&i2c, speeds[i]));
-    EXECUTE_TEST(test_result, write_read_random);
-    EXECUTE_TEST(test_result, write_read_page);
+  status_t test_result = OK_STATUS();
+  dif_i2c_t i2c;
+  i2c_pinmux_platform_id_t platform = I2cPinmuxPlatformIdCw310Pmod;
+  for (uint8_t i2c_instance = 0; i2c_instance < 3; ++i2c_instance) {
+    LOG_INFO("Testing i2c%d", i2c_instance);
+    CHECK_STATUS_OK(test_init(&i2c, i2c_instance, platform));
+
+    dif_i2c_speed_t kSpeeds[] = {kDifI2cSpeedStandard, kDifI2cSpeedFast,
+                                 kDifI2cSpeedFastPlus};
+
+    for (size_t i = 0; i < ARRAYSIZE(kSpeeds); ++i) {
+      CHECK_STATUS_OK(i2c_testutils_set_speed(&i2c, kSpeeds[i]));
+      EXECUTE_TEST(test_result, write_read_random, &i2c);
+      EXECUTE_TEST(test_result, write_read_page, &i2c);
+    }
```

Comment:
Do we also need to disable the i2c instance when we're done with it? (and possibly put its peripheral input muxes in a tied-high state?)

Reply:
I changed to disable the i2c instance. 
But I don't understand what you mean about the "peripheral input muxes in a tied-high state".

Reply:
The SCA and SCL inputs of the particular i2c instance are driven by the output of corresponding peripheral input muxes (in the pinmux). We have two types of muxed in the pinmux:

- peripheral input muxes
    - Configures what a peripheral sees on its inputs
- pad output muxes
    - Configures what a pad drives on its output

I think the inputs of the disabled i2c are being allowed to toggle here (still connected in the pinmux). That should be fine functionally, but that probably would not be what typically happens in practice, hehe.

Reply:
I see, so if I understood correctly at the end we would have all the i2c instances connected to the bus?

Reply:
> I see, so if I understood correctly at the end we would have all the i2c instances connected to the bus?

That's halfway correct, yes. All i2c _inputs_ are connected to the bus. All the *_sda_i and *_scl_i inputs are receiving from the same set of MIO pads. The IOB11 and IOB12 MIO pad drivers, however, get changed away from the last set of i2cX_sda_o and i2cX_scl_o outputs with each `i2c_testutils_select_pinmux()` call.

> I can see that the instance insel can be muxed to `kTopEarlgreyPinmuxInselConstant[One|Zero]` but there's nothing similar for the mio_out.

The `mio_out` stuff is for pad driver values, and you're taking care of those muxes with `dif_pinmux_output_select`.

Muxes are best identified by the 1 thing their output drives. Pad driver muxes drive the MIO pads:
```
i2c0_sda0_o---|  \
              |   ----- IOA0 (driver)
i2c1_sda0_o___|  /

i2c0_sda0_o---|  \
              |   ----- IOA1 (driver)
i2c1_sda0_o___|  /
    .
    .
    .
```

Those are handled by `dif_pinmux_output_select`.

On the other hand, there is a mux driving each _peripheral input_:
```
IOA0 (receiver)---|  \
                  |   ----- i2c0_sda0_i (I2C0's SDA input)
IOA1 (receiver)---|  /
                  | /
...               |/


IOA0 (receiver)---|  \
                  |   ----- i2c1_sda0_i (I2C1's SDA input)
IOA1 (receiver)---|  /
                  | /
...               |/
```

Those are handled by `dif_pinmux_input_select`, and the mux driving I2C0's SDA input is independent of the mux driving I2C1's SDA input. It's _these_ muxes that don't reset to a quiescent state (ConstantZero or ConstantOne) with each call to `pinmux_testutils_select_pinmux()`.

Reply:
This setup, btw, makes our pinmux unusual. Typically, MCUs do not have independent controls for inputs and outputs. Unlike those, our pinmux technically lets you do odd configurations, like...
```
IOA0            -> i2c0.sda.input

i2c0.sda.output -> IOB13
```